src/templates/Async.java

```diff
@@ -0,0 +1,146 @@
+package {{ config.package_name() }};
+
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.function.BiFunction;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+public final class UniffiAsyncHelpers {
+    // Async return type handlers
+    static final byte UNIFFI_RUST_FUTURE_POLL_READY = (byte) 0;
+    static final byte UNIFFI_RUST_FUTURE_POLL_MAYBE_READY = (byte) 1;
+    static final UniffiHandleMap<CompletableFuture<Byte>> uniffiContinuationHandleMap = new UniffiHandleMap<>();
+
+    // FFI type for Rust future continuations
+    enum UniffiRustFutureContinuationCallbackImpl implements UniffiRustFutureContinuationCallback {
+        INSTANCE;
+
+        @Override
+        public void callback(Long data, Byte pollResult) {
+            uniffiContinuationHandleMap.remove(data).complete(pollResult);
+        }
+    }
+
+    @FunctionalInterface
+    interface PollingFunction {
+        void apply(long rustFuture, UniffiRustFutureContinuationCallback callback, long continuationHandle);
+    }
+
+    static <T, F, E extends Exception> CompletableFuture<T> uniffiRustCallAsync(
+        long rustFuture,
+        PollingFunction pollFunc,
+        BiFunction<Long, UniffiRustCallStatus, F> completeFunc,
+        Consumer<Long> freeFunc,
+        Function<F, T> liftFunc,
+        UniffiRustCallStatusErrorHandler<E> errorHandler
+    ){
+        CompletableFuture<T> future = new CompletableFuture<>();
+
+        // TODO(murph): may want an overload that takes an executor to run on.
```

Comment:
I agree that it might be misleading, but how do we expect the caller to put the futures on an blocking thread pool otherwise? We could just have one, but then they can't control how many things are allowed to be blocked.

Reply:
I don't think there's a good way to thread an executor down to this point to be honest. This function call is made by generated code. It seems like other clients in Java with CFs either don't allow you to set the executor or set it on the client itself, not the individual calls. Java's internal HTTP client does expose `_Async` versions of all the methods that take an executor, if we wanted to support it I think that's what we'd have to do. Double the API surface by generating both everywhere we expose an async method.

I'll make a ticket for that.

Reply:
https://github.com/IronCoreLabs/uniffi-bindgen-java/issues/13